### PR TITLE
CGE-68: Refactoring exception handlers

### PR DESCRIPTION
**Pull request by KozminAlexandr**
All exception handlers were moved from REST controllers.
All exception handlers were covered by unit tests.

### CODE REVIEW

File: src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                messageSourceUtil.getExceptionMessageSource("illegal.argument.exception"),
+                e.getLocalizedMessage(), request.getRequest().getRequestURI(), request.getHttpMethod());
+
+        log.warn(errorMessage.errorLog());
+
+        return new ResponseEntity<>(errorMessage, HttpStatus.BAD_REQUEST);
+    }
+
+    @ExceptionHandler(ConstraintViolationException.class)
+    public ResponseEntity<Object> handleConstraintViolationException(ConstraintViolationException e,
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java:122]: Could you remind me when we get this one exception?

File: src/main/resources/messages.properties [hunk 2]
```diff
@@ -17,23 +17,19 @@ token.expired=JWT expired at
 
 invalid.category.id=There is no category by id: 
 category.not-deletable=The category can not be deleted by this id, because it has to exist by id and it's subcategories mustn't have products! Given category id: 
-invalid.new-category-dto=New category or subcategory mustn't contain id except 0, also category name mustn't have duplicates!";
+invalid.new-category-dto=New category or subcategory mustn't contain id except 0, also category name mustn't have duplicates!
```
Comment (rpkyrych) [inline src/main/resources/messages.properties:20]: Thank for this! :+1:

File: src/main/resources/messages.properties [hunk 3]
```diff
@@ -42,4 +38,14 @@ too.big.email=Email length should be less than 129 characters. Please enter vali
 
 valid.email.regexp=^[\\w-\\+]+(\\.[\\w]+)*@[\\w-]+(\\.[\\w]+)*(\\.[a-zA-Z]{2,})$
 valid.password.regexp=(?=.*?[0-9])(?=.*?[a-z])(?=.*?[A-Z])(?=\\S+$).+.{8,30}
-valid.username.regexp=(?=\\S+$).{2,50}
\ No newline at end of file
+valid.username.regexp=(?=\\S+$).{2,50}
+
+## Exceptions messages
```
Comment (rpkyrych) [inline src/main/resources/messages.properties:43]: Rename to "Exception messages" ;)

File: src/main/resources/messages.properties [hunk 4]
```diff
@@ -42,4 +38,14 @@ too.big.email=Email length should be less than 129 characters. Please enter vali
...
+user.not-found.exception=User not found
+security.exception=Security breach
+illegal.id.exception=Identifier does not match
+invalid.dto.exception=Invalid data transfer object
+sql.exception=Exception during saving object to the database
+entity.not-found.exception=Entity not found into the database
+illegal.argument.exception=Exception during request. Illegal or inappropriate argument
+validation.exception=Validation exception
```
Comment (rpkyrych) [inline src/main/resources/messages.properties:51]: Make these parameter more consistent - move word "exception" forward:
**exception.user.not-found;
exception.illegal.id.**

File: src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
+package com.hillel.items_exchange.exceptionshandling;
```
Comment (rpkyrych) [inline src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java:1]: Move current class to "exception.handler" package.

File: src/test/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        mockMvc = MockMvcBuilders.standaloneSetup(advertisementController, categoryController)
+                .setControllerAdvice(new GlobalControllerExceptionHandler(messageSourceUtil))
+                .build();
+    }
+
+    @Test
+    public void testHandleUserNotFoundException() throws Exception {
+        when(advertisementController.createAdvertisement(any(), any())).thenThrow(UsernameNotFoundException.class);
+
+        MvcResult result = mockMvc.perform(post("/adv")
```
Comment (rpkyrych) [inline src/test/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandlerTest.java:74]: This part could be refactored as well (put, post):
`getResult(MockHttpServletRequestBuilder requestType, AdvertisementDto dto, ResultMatcher result) throws Exception`

File: src/test/resources/messages.properties [hunk 7]
```diff
@@ -41,4 +39,14 @@ too.big.email=Email length should be less than 129 characters. Please enter vali
 
 valid.email.regexp=^[\\w-\\+]+(\\.[\\w]+)*@[\\w-]+(\\.[\\w]+)*(\\.[a-zA-Z]{2,})$
 valid.password.regexp=(?=.*?[0-9])(?=.*?[a-z])(?=.*?[A-Z])(?=\\S+$).+.{8,30}
-valid.username.regexp=(?=\\S+$).{2,50}
\ No newline at end of file
+valid.username.regexp=(?=\\S+$).{2,50}
+
+## Exceptions messages
```
Comment (rpkyrych) [inline src/test/resources/messages.properties:44]: Apply here my comments from the same file but for prod.

File: src/main/java/com/hillel/items_exchange/controller/SubcategoryController.java [hunk 8]
```diff
@@ -47,20 +46,8 @@
...
-
-    @ExceptionHandler(SQLIntegrityConstraintViolationException.class)
-    public ResponseEntity<String> handleSqlException(SQLIntegrityConstraintViolationException e) {
-        log.error(e.getMessage(), e);
-        return ResponseEntity
-                .status(HttpStatus.BAD_REQUEST)
-                .body(messageSourceUtil.getExceptionMessageSourceWithAdditionalInfo("sql.exception",
-                        e.getLocalizedMessage()));
+    @GetMapping(value = "/exist/{subcategory_id}")
+    public boolean isSubcategoryExistsById(@PathVariable("subcategory_id") long id) {
```
[the remark this answers is not in the document]
Comment (KozminAlexandr) [reply src/main/java/com/hillel/items_exchange/controller/SubcategoryController.java:50]: It is my incorrect merge CGE-45 branch to CGE-68. I already fixed this. You are very observant =)  Thanks!

File: src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java (hunk 1 above)
Comment (KozminAlexandr) [reply src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java]: When JSR-303 validation failed

File: src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+import javax.validation.ConstraintViolation;
+import javax.validation.ConstraintViolationException;
+import java.util.Date;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@ControllerAdvice
+@Log4j2
+@AllArgsConstructor
+public class GlobalControllerExceptionHandler {
```
[the remark this answers is not in the document]
Comment (KozminAlexandr) [reply src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java]: No, by the way, about Swagger.. now responces from @ControllerAdvice is undocumented. Do I resolve this in this task, or for this needs another?

File: src/main/java/com/hillel/items_exchange/exception/handler/GlobalExceptionHandler.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+            .withLocale(Locale.getDefault());
+
+    private final MessageSourceUtil messageSourceUtil;
+
+    @ExceptionHandler(UsernameNotFoundException.class)
+    public ResponseEntity<ErrorMessage> handleUserNotFoundException(UsernameNotFoundException e,
+                                                                    ServletWebRequest request) {
+
+        ErrorMessage errorMessage = getErrorMessage(request, HttpStatus.NOT_FOUND,
+                "exception.user.not-found", Collections.singletonList(e.getLocalizedMessage()));
```
Comment (KozminAlexandr) [inline src/main/java/com/hillel/items_exchange/exception/handler/GlobalExceptionHandler.java:46]: In the previous version there was an option with `List.of`. But to `List.of` cannot add `null`. Сhance of this approaching zero, but...

File: src/main/java/com/hillel/items_exchange/exceptionshandling/ErrorMessage.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    public ErrorMessage(Date timestamp, int status, String error, List<String> details, String path, HttpMethod method) {
+        this.timestamp = dateFormat.format(timestamp);
+        this.status = status;
+        this.error = error;
+        this.details = details;
+        this.path = path;
+        this.method = method;
+    }
+
+    public ErrorMessage(Date timestamp, int status, String error, String detail, String path, HttpMethod method) {
```
[the remark this answers is not in the document]
Comment (KozminAlexandr) [reply src/main/java/com/hillel/items_exchange/exceptionshandling/ErrorMessage.java]: I use just @AllArgsConstructor, after refactoring ErrorMessage it turned out enough

File: src/main/java/com/hillel/items_exchange/exceptionshandling/ErrorMessage.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        this.timestamp = dateFormat.format(timestamp);
+        this.status = status;
+        this.error = error;
+        this.details = Collections.singletonList(detail);
+        this.path = path;
+        this.method = method;
+    }
+
+    public StringBuilder errorLog() {
+        return new StringBuilder("Api error: ")
```
[the remark this answers is not in the document]
Comment (KozminAlexandr) [reply src/main/java/com/hillel/items_exchange/exceptionshandling/ErrorMessage.java]: Performed. Now it looks much better =)

File: src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+        log.warn(errorMessage.errorLog());
+
+        return new ResponseEntity<>(errorMessage, HttpStatus.NOT_FOUND);
+    }
+
+    @ExceptionHandler(IllegalIdentifierException.class)
+    public ResponseEntity<Object> handleIllegalIdentifierException(IllegalIdentifierException e,
+                                                                   ServletWebRequest request) {
+
+        ErrorMessage errorMessage = new ErrorMessage(new Date(), HttpStatus.BAD_REQUEST.value(),
```
[the remark this answers is not in the document]
Comment (KozminAlexandr) [reply src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java]: Performed

File: src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+                messageSourceUtil.getExceptionMessageSource("illegal.id.exception"),
+                e.getMessage(), request.getRequest().getRequestURI(), request.getHttpMethod());
+
+        log.warn(errorMessage.errorLog());
```
[the remark this answers is not in the document]
Comment (KozminAlexandr) [reply src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java]: Yes, this message saves in LogFile.log. Moved to new method

File: src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+                messageSourceUtil.getExceptionMessageSource("entity.not-found.exception"),
+                e.getLocalizedMessage(), request.getRequest().getRequestURI(), request.getHttpMethod());
+
+        log.warn(errorMessage.errorLog());
+
+        return new ResponseEntity<>(errorMessage, HttpStatus.NOT_FOUND);
+    }
+
+    @ExceptionHandler(IllegalIdentifierException.class)
+    public ResponseEntity<Object> handleIllegalIdentifierException(IllegalIdentifierException e,
```
[the remark this answers is not in the document]
Comment (KozminAlexandr) [reply src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java]: I don’t even know what I was thinking =) Changed to ErrorMessage

File: src/main/resources/messages.properties (hunk 3 above)
Comment (KozminAlexandr) [reply src/main/resources/messages.properties]: Performed

File: src/main/resources/messages.properties (hunk 4 above)
Comment (KozminAlexandr) [reply src/main/resources/messages.properties]: Performed

File: src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java (hunk 5 above)
Comment (KozminAlexandr) [reply src/main/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandler.java]: Performed

File: src/test/java/com/hillel/items_exchange/controller/ImageControllerTest.java [hunk 16]
```diff
@@ -9,13 +9,12 @@
...
 import org.springframework.test.web.servlet.MockMvc;
 
-
 import javax.transaction.Transactional;
 
 import static org.hamcrest.Matchers.hasSize;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
-import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
 import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
```
Comment (rpkyrych) [inline src/test/java/com/hillel/items_exchange/controller/ImageControllerTest.java:17]: You shouldn't create such type commits in the future.

File: src/test/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandlerTest.java (hunk 6 above)
Comment (KozminAlexandr) [reply src/test/java/com/hillel/items_exchange/exceptionshandling/GlobalControllerExceptionHandlerTest.java]: I did a little different and now we can use `getResult` method not only for put and post http methods

File: src/test/resources/messages.properties (hunk 7 above)
Comment (KozminAlexandr) [reply src/test/resources/messages.properties]: Performed

File: src/test/java/com/hillel/items_exchange/exceptionshandling/GlobalExceptionHandlerTest.java [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+
+        MvcResult result = mockMvc.perform(get("/image/{product_id}", -1L)
+                .accept(MediaType.APPLICATION_JSON))
+                .andExpect(status().isBadRequest())
+                .andReturn();
+
+        assertThat(result.getResolvedException(), is(instanceOf(ConstraintViolationException.class)));
+    }
+
+    private MvcResult getResult(HttpMethod httpMethod, String path, AdvertisementDto dto,
```
Comment (rpkyrych) [inline src/test/java/com/hillel/items_exchange/exceptionshandling/GlobalExceptionHandlerTest.java:147]: Looks much better! :+1:

**Pull request merged**